velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp

```diff
@@ -2201,5 +2201,142 @@ TEST_F(MinMaxByNTest, incrementalWindow) {
   }
 }
 
+TEST_F(MinMaxByComplexTypes, arrayTypeCompare) {
+  // The data consists of five rows, and the comparison key is arrayType.
+  // Each array contains two elements.
+  auto data = makeRowVector({
+      makeFlatVector<int64_t>({8, 3, 4, 5, 6}),
+      makeArrayVector<int64_t>({{3, 4}, {2, 4}, {2, 3}, {0, 3}, {0, 2}}),
+  });
+
+  auto expected = makeRowVector({makeArrayVector<int64_t>({
+      {6, 5},
+  })});
+
+  testAggregations({data}, {}, {"min_by(c0, c1, 2)"}, {expected});
+
+  // The data consists of five rows, and the comparison key is arrayType.
+  // Each array contains two elements.
+  data = makeRowVector({
+      makeFlatVector<int16_t>({1, 2, 1, 2, 1}),
+      makeFlatVector<int64_t>({8, 3, 4, 5, 6}),
+      makeArrayVector<int64_t>({{3, 4}, {2, 4}, {2, 3}, {0, 3}, {0, 2}}),
+  });
+
+  expected = makeRowVector(
+      {makeFlatVector<int16_t>({1, 2}),
+       makeArrayVector<int64_t>({{8, 4}, {3, 5}})});
+
+  testAggregations({data}, {"c0"}, {"max_by(c1, c2, 2)"}, {expected});
+}
+
+TEST_F(MinMaxByComplexTypes, mapTypeCompare) {
```

Comment:
Can you add a test with nulls in map's/arrays (and null arrays) and also validate that behavior with Presto ?

Reply:
Thanks for this. I found something when validating the behavior with Presto.

1. For Array type, Presto supports array without nulls.

Without nulls.
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 43.0, ARRAY[10, 30]),
     ->     (2, 43.0, ARRAY[20, 10]),
     ->     (3, 53.0, ARRAY[50, 40])
     ->   ) AS t (id, col1, col2)
     -> ) SELECT min_by(col1, col2, 2)
     -> FROM dataset;
    _col0     
--------------
 [43.0, 43.0] 
```

With nulls.
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 43.0, ARRAY[10, NULL, 30]),
     ->     (2, 43.0, ARRAY[20, NULL, 10]),
     ->     (3, 53.0, ARRAY[NULL, 50, 40])
     ->   ) AS t (id, col1, col2)
     -> ) SELECT min_by(col1, col2, 2)
     -> FROM dataset;
Query 20240321_092745_00006_j8s5b, FAILED, 1 node
Splits: 6 total, 0 done (0.00%)
[Latency: client-side: 107ms, server-side: 83ms] [0 rows, 0B] [0 rows/s, 0B/s]

Query 20240321_092745_00006_j8s5b failed: ARRAY comparison not supported for arrays with null elements
```
2. Presto not support MapType as key(also test in demo ut).
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 43.0, MAP(ARRAY['a', 'b'], ARRAY[10, 11])),
     ->     (2, 43.0, MAP(ARRAY['c', 'd'], ARRAY[11, 10])),
     ->     (3, 53.0, MAP(ARRAY['e', 'f'], ARRAY[50, 11]))
     ->   ) AS t (id, col1, col2)
     -> ) SELECT
     ->   min_by(col1, col2, 2)
     -> FROM dataset;
Query 20240321_093011_00007_j8s5b failed: line 8:3: 
Unexpected parameters (decimal(3,1), map(varchar(1),integer), integer) for function min_by. Expected: min_by(V, K) K:orderable, V, min_by(V, K, bigint) V, K:orderable
```
3. Presto support RowType as key, but not support RowType col **which participate in comparison** contains null
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 42.0, ROW(10, 'A', TRUE)),
     ->     (2, 43.0, ROW(20, 'B', FALSE)),
     ->     (3, 53.0, ROW(50, 'C', TRUE))
     ->   ) AS t (id, col1, col2)
     -> ) 
     -> SELECT
     ->   min_by(col1, col2, 2)
     -> FROM dataset;
    _col0     
--------------
 [42.0, 43.0] 
(1 row)
```

The second col does not participate in comparison.
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 42.0, ROW(10, NULL, TRUE)),
     ->     (2, 43.0, ROW(20, 'B', FALSE)),
     ->     (3, 53.0, ROW(50, 'C', TRUE))
     ->   ) AS t (id, col1, col2)
     -> ) 
     -> SELECT
     ->   min_by(col1, col2, 2)
     -> FROM dataset;
    _col0     
--------------
 [42.0, 43.0] 
(1 row)
```

The second col participate in comparison.
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 42.0, ROW(10, NULL, TRUE)),
     ->     (2, 43.0, ROW(10, 'B', FALSE)),
     ->     (3, 53.0, ROW(10, 'C', TRUE))
     ->   ) AS t (id, col1, col2)
     -> ) 
     -> SELECT
     ->   min_by(col1, col2, 2)
     -> FROM dataset;

Query 20240321_094004_00016_j8s5b, FAILED, 1 node
Splits: 6 total, 0 done (0.00%)
[Latency: client-side: 101ms, server-side: 72ms] [0 rows, 0B] [0 rows/s, 0B/s]

Query 20240321_094004_00016_j8s5b failed: ROW comparison not supported for fields with null elements
```

Reply:
So I will look at how to validate the behavior with Presto.

remove mapType in this PR
Throw an exception when the key col which participate in comparison contains null.
How do you think?